src/server.rs

```diff
@@ -0,0 +1,162 @@
+use hyper::{self, body};
+use hyper::{server::conn::http1, service::service_fn};
+use hyper::{Method, Request, Response, StatusCode};
+use hyper_util::rt::tokio::TokioIo;
+
+use http_body_util::Full;
+
+use bytes::Bytes;
+
+use tokio::net::TcpListener;
+use tokio_util::sync::CancellationToken;
+
+use std::collections::HashMap;
+use std::net::SocketAddr;
+
+use tokio::sync::{mpsc, oneshot};
+
+use serde::{Deserialize, Serialize};
+
+use crate::metrics;
+
+use log::*;
+
+fn parse_query(q: &str) -> HashMap<String, String> {
+    form_urlencoded::parse(q.as_bytes())
+        .into_owned()
+        .collect::<HashMap<String, String>>()
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SummaryStats {
+    min: u64,
+    max: u64,
+    median: u64,
+}
+
+#[derive(Debug, Clone)]
+pub struct MetricsService {
+    m: mpsc::Sender<metrics::MetricsCommands>,
+}
+
+impl MetricsService {
+    pub fn new(m: mpsc::Sender<metrics::MetricsCommands>) -> MetricsService {
+        MetricsService { m }
+    }
+
+    async fn get_summary_stats(
+        &mut self,
+        from: Option<std::time::Duration>,
+    ) -> Option<SummaryStats> {
+        let (tx, rx) = oneshot::channel();
+
+        if let Err(e) = self
+            .m
+            .send(metrics::MetricsCommands::GetLastMetrics {
+                window_duration: from,
+                resp: tx,
+            })
+            .await
+        {
+            error!("Couldn't get summary stats: {}", e);
+            return None;
+        }
+
+        match rx.await {
+            Ok(resp) => Some(SummaryStats {
+                min: resp.hist.percentile(0.0).unwrap().end(),
+                max: resp.hist.percentile(100.0).unwrap().end(),
+                median: resp.hist.percentile(50.0).unwrap().start(),
```

Comment:
This metric selection is a bit sparse?

Reply:
Yeah, for now the only metrics generated are the bare minimum ducktape will likely need to determine if the clients have made progress. The framework to add more is there and it should be pretty easy to add more in the future as the need arises.